tests/Microsoft.Identity.Test.Unit.net45/CacheTests/TokenCacheNotificationTests.cs

```diff
@@ -212,5 +213,61 @@ public async Task TestSerializationViaAsync()
+                    .Create(TestConstants.ClientId)
+                    .WithHttpManager(_harness.HttpManager)
+                    .BuildConcrete();
+
+                Trace.WriteLine("Step 1 - call GetAccounts with empty cache - no accounts, no tokens");
+                var cacheAccessRecorder1 = app.UserTokenCache.RecordAccess();
+                var accounts = await app.GetAccountsAsync().ConfigureAwait(false);
+
+                Assert.IsFalse(cacheAccessRecorder1.LastBeforeAccessNotificationArgs.HasTokens);
+                Assert.IsNull(cacheAccessRecorder1.LastBeforeWriteNotificationArgs);
+                Assert.IsFalse(cacheAccessRecorder1.LastAfterAccessNotificationArgs.HasTokens);
+                Assert.IsFalse(accounts.Any());
+
+                Trace.WriteLine("Step 2 - call AcquireTokenInteractive - it will save new tokens in the cache");
+                MsalMockHelpers.ConfigureMockWebUI(
+                        app.ServiceBundle.PlatformProxy,
+                         AuthorizationResult.FromUri(app.AppConfig.RedirectUri + "?code=some-code"));
+                _harness.HttpManager.AddInstanceDiscoveryMockHandler();
+                _harness.HttpManager.AddSuccessTokenResponseMockHandlerForPost();
+                var cacheAccessRecorder2 = app.UserTokenCache.RecordAccess();
+
+                await app
+                    .AcquireTokenInteractive(TestConstants.s_scope)
+                    .ExecuteAsync()
+                    .ConfigureAwait(false);
+
+                Assert.IsFalse(cacheAccessRecorder2.LastBeforeAccessNotificationArgs.HasTokens);
+                Assert.IsFalse(cacheAccessRecorder2.LastBeforeWriteNotificationArgs.HasTokens);
+                Assert.IsTrue(cacheAccessRecorder2.LastAfterAccessNotificationArgs.HasTokens);
+
+                Trace.WriteLine("Step 3 - call GetAccounts - now with 1 account");
+                var cacheAccessRecorder3 = app.UserTokenCache.RecordAccess();
+                accounts = await app.GetAccountsAsync().ConfigureAwait(false);
+
+                Assert.IsTrue(cacheAccessRecorder3.LastBeforeAccessNotificationArgs.HasTokens);
+                Assert.IsNull(cacheAccessRecorder3.LastBeforeWriteNotificationArgs);
+                Assert.IsTrue(cacheAccessRecorder3.LastAfterAccessNotificationArgs.HasTokens);
+                Assert.IsTrue(accounts.Any());
+
+                Trace.WriteLine("Step 4 - call RemoveAccounts - this will delete all the tokens");
+                var cacheAccessRecorder4 = app.UserTokenCache.RecordAccess();
+                await app.RemoveAsync(accounts.Single()).ConfigureAwait(false);
+
+                Assert.IsTrue(cacheAccessRecorder4.LastBeforeAccessNotificationArgs.HasTokens);
+                Assert.IsTrue(cacheAccessRecorder4.LastBeforeWriteNotificationArgs.HasTokens);
+                Assert.IsFalse(cacheAccessRecorder4.LastAfterAccessNotificationArgs.HasTokens);
+            }
```

Comment:
@jmprieur @TiagoBrenck - please let me know if this behavior of the flag is intuitive enough. The flag tells you if the in-memory cache has tokens, so the Before* callbacks are not really interesting because it does not matter if the current in-memory cache has tokens or not, tokens can be injected from the persisted cache.

Reply:
@bgavrilMS : I think it's well explained in the XML comment. We'll use it in Microsoft.Identity.Web and that will do (this is mostly for Confidential client apps)